diff -urN pidentd-3.1a17.org/Makefile.in pidentd-3.1a17/Makefile.in
--- pidentd-3.1a17.org/Makefile.in	Fri Nov 17 12:21:21 2000
+++ pidentd-3.1a17/Makefile.in	Fri Nov 17 12:21:52 2000
@@ -28,8 +28,8 @@
 
 install:
 	@cd src ; $(MAKE) CC="$(CC)" CFLAGS="$(MY_CFLAGS)" LDFLAGS="$(LDFLAGS)" install
-	-mkdir -p $(mandir)/man8
-	$(INSTALL) -m 644 $(srcdir)/doc/identd.8 $(mandir)/man8
+	-mkdir -p $(DESTDIR)$(mandir)/man8
+	$(INSTALL) -m 644 $(srcdir)/doc/identd.8 $(DESTDIR)$(mandir)/man8
 
 
 clean mostlyclean distclean maintainer-clean dist sign upload version:
diff -urN pidentd-3.1a17.org/plib/Makefile.in pidentd-3.1a17/plib/Makefile.in
--- pidentd-3.1a17.org/plib/Makefile.in	Fri Nov 17 12:21:21 2000
+++ pidentd-3.1a17/plib/Makefile.in	Fri Nov 17 12:22:49 2000
@@ -36,9 +36,9 @@
 # Installing
 
 install: $(LIBS)
-	-mkdir -p $(libdir)
+	-mkdir -p $(DESTDIR)$(libdir)
 	for P in $(LIBS) ; do \
-		$(INSTALL) -m 644 $${P} $(libdir); \
+		$(INSTALL) -m 644 $${P} $(DESTDIR)$(libdir); \
 	done
 
 
diff -urN pidentd-3.1a17.org/src/Makefile.in pidentd-3.1a17/src/Makefile.in
--- pidentd-3.1a17.org/src/Makefile.in	Fri Nov 17 12:21:21 2000
+++ pidentd-3.1a17/src/Makefile.in	Fri Nov 17 12:22:34 2000
@@ -66,9 +66,9 @@
 # Installing
 
 install:	$(PROGRAMS)
-	-mkdir -p $(sbindir)
+	-mkdir -p $(DESTDIR)$(sbindir)
 	for P in $(PROGRAMS) ; do \
-		$(INSTALL) -m 755 $${P} $(sbindir); \
+		$(INSTALL) -m 755 $${P} $(DESTDIR)$(sbindir); \
 	done
 #	-mkdir -p $(etcdir)
 #	$(INSTALL) -m 644 identd.conf $(etcdir)
